pkg/base/helm.go

```diff
@@ -271,3 +274,105 @@ func checkChartForVersion(file *upstreamtypes.UpstreamFile) (string, error) {
 	// if no determination is made, assume v2
 	return "v2", nil
 }
+
+// insert namespace if it's defined in the spec and not already present in the manifests
+func kustomizeHelmNamespace(baseFiles []BaseFile, renderOptions *RenderOptions) ([]BaseFile, error) {
+	var updatedBaseFiles []BaseFile
+	if renderOptions.Namespace != "" {
+		chartsPath, err := ioutil.TempDir("", "charts")
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create temp dir")
+		}
+		defer os.RemoveAll(chartsPath)
+
+		resources := map[string]BaseFile{}
+		var kustomizeResources []string
+		var kustomizePatches []kustomizetypes.PatchStrategicMerge
+		foundGVKNamesMap := map[string]bool{}
+		for _, baseFile := range baseFiles {
+			// write temp files for manifests that need a namespace
+			gvk, manifest := GetGVKWithNameAndNs(baseFile.Content, renderOptions.Namespace)
+			if manifest.APIVersion == "" && manifest.Kind == "" && manifest.Metadata.Name == "" {
+				updatedBaseFiles = append(updatedBaseFiles, baseFile)
+				continue // ignore invalid resources
+			}
+
+			if manifest.Metadata.Namespace == "" {
+				splitPath := strings.Split(baseFile.Path, "/")
+				name := splitPath[len(splitPath)-1]
+				tmpFile, err := ioutil.TempFile(chartsPath, name)
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to write temp file %v", tmpFile.Name())
+				}
+
+				if _, err := tmpFile.Write(baseFile.Content); err != nil {
+					return nil, errors.Wrapf(err, "failed to write temp file %v content", tmpFile.Name())
+				}
+				if err := tmpFile.Close(); err != nil {
+					return nil, errors.Wrapf(err, "failed to close temp file %v", tmpFile.Name())
+				}
+
+				if !foundGVKNamesMap[gvk] || gvk == "" {
```

Comment:
since there are potentially duplicates (i.e. patches), i still had to filter those and add them to `kustomization.yaml` correctly. that means those patches are already applied so i chose to not pass them on to the next step. let me know if we prefer passing them on anyway. for resources that have a hardcoded namespace, duplicates are maintained and added to base like before.